lib/active_merchant/billing/gateways/decidir.rb

```diff
@@ -167,29 +167,55 @@ def add_amount(post, money, options)
         post[:amount] = localized_amount(money, currency).to_i
       end
 
-      def add_payment(post, credit_card, options)
-        card_data = {}
+      def add_payment(post, payment_method, options)
+        add_common_payment_data(post, payment_method, options)
+
+        case payment_method
+        when NetworkTokenizationCreditCard
+          add_network_token(post, payment_method, options)
+        else
+          add_credit_card(post, payment_method, options)
+        end
+      end
+
+      def add_common_payment_data(post, payment_method, options)
+        post[:card_data] = {}
+
+        data = post[:card_data]
+        data[:card_holder_identification] = {}
+        data[:card_holder_identification][:type] = options[:card_holder_identification_type] if options[:card_holder_identification_type]
+        data[:card_holder_identification][:number] = options[:card_holder_identification_number] if options[:card_holder_identification_number]
+        data[:card_holder_name] = payment_method.name if payment_method.name
+
+        # additional data used for Visa transactions
+        data[:card_holder_door_number] = options[:card_holder_door_number].to_i if options[:card_holder_door_number]
+        data[:card_holder_birthday] = options[:card_holder_birthday] if options[:card_holder_birthday]
+      end
+
+      def add_network_token(post, payment_method, options)
+        post[:is_tokenized_payment] = true
+        post[:fraud_detection] ||= {}
+        post[:fraud_detection][:sent_to_cs] = false
+        post[:card_data][:last_four_digits] = payment_method.last_digits
+
+        post[:token_card_data] = {
+          token: payment_method.number,
+          eci: payment_method.eci,
```

Comment:
Is the `eci` value required for Decidir? It's not required for network tokenization for all gateways. If not, maybe there should be a check for the presence of `eci` value on the `payment_method` before populating the `eci` field.

Reply:
Yes, it's required.